File: flink-code-scala-0421/src/main/scala/com/atguigu/day02/FlatMapExample.scala
package com.atguigu.day02

import com.atguigu.day02.util.SensorReading
import org.apache.flink.api.common.functions.FlatMapFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

object FlatMapExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream = env.fromElements("white", "black", "gray")

    stream
      .flatMap(new FlatMapFunction[String, String] {
        override def flatMap(t: String, collector: Collector[String]): Unit = {
          if (t.equals("white")) {
            collector.collect(t)
          } else if (t.equals("black")) {
            collector.collect(t)
            collector.collect(t)
          }
        }
      })

    stream
      .flatMap(new MyFlatMap)
      .print()

    env.execute()
  }

  class MyFlatMap extends FlatMapFunction[String, String] {
    override def flatMap(t: String, collector: Collector[String]): Unit = {
      if (t.equals("white")) {
        collector.collect(t)
      } else if (t.equals("black")) {
        collector.collect(t)
        collector.collect(t)
      }
    }
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day02/FlatMapExample.java
package com.atguigu.day02;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

public class FlatMapExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<String> stream = env.fromElements("white", "black", "gray");

        stream
                .flatMap(new FlatMapFunction<String, String>() {
                    @Override
                    public void flatMap(String s, Collector<String> collector) throws Exception {
                        if (s.equals("white")) {
                            collector.collect(s);
                        } else if (s.equals("black")) {
                            collector.collect(s);
                            collector.collect(s);
                        }
                    }
                })
                .print();

        stream
                .flatMap(new MyFlatMap())
                .print();

        env.execute();
    }

    public static class MyFlatMap implements FlatMapFunction<String, String> {
        @Override
        public void flatMap(String s, Collector<String> collector) throws Exception {
            if (s.equals("white")) {
                collector.collect(s);
            } else if (s.equals("black")) {
                collector.collect(s);
                collector.collect(s);
            }
        }
    }
}
